File: benchmarks/rust/mandelbrot.rs
// rustc -O mandelbrot.rs -o mandelbrot_rs

fn mandel_iter(cr: f64, ci: f64, max_iter: i64) -> i64 {
    let mut zr = 0.0_f64;
    let mut zi = 0.0_f64;
    let mut i: i64 = 0;
    while i < max_iter {
        let zr2 = zr * zr;
        let zi2 = zi * zi;
        if zr2 + zi2 > 4.0 {
            return i;
        }
        let new_zr = zr2 - zi2 + cr;
        let new_zi = 2.0 * zr * zi + ci;
        zr = new_zr;
        zi = new_zi;
        i += 1;
    }
    max_iter
}

fn main() {
    let width: i64 = 200;
    let height: i64 = 200;
    let max_iter: i64 = 1000;
    let mut checksum: i64 = 0;
    let mut y: i64 = 0;
    while y < height {
        let mut x: i64 = 0;
        while x < width {
            let cr = -2.0 + 3.0 * (x as f64) / (width as f64);
            let ci = -1.5 + 3.0 * (y as f64) / (height as f64);
            let it = mandel_iter(cr, ci, max_iter);
            checksum += it;
            x += 1;
        }
        y += 1;
    }
    println!("mandelbrot checksum = {}", checksum);
}

File: benchmarks/mandelbrot.c
#include <stdio.h>
typedef long i64;
static i64 mandel(i64 cr, i64 ci, i64 max_iter, i64 scale) {
    i64 zr = 0, zi = 0, i = 0;
    i64 four = 4 * scale * scale;
    while (i < max_iter) {
        i64 zr2 = (zr * zr) / scale;
        i64 zi2 = (zi * zi) / scale;
        if (zr2 + zi2 > four) return i;
        i64 new_zr = zr2 - zi2 + cr;
        i64 new_zi = (2 * zr * zi) / scale + ci;
        zr = new_zr;
        zi = new_zi;
        i++;
    }
    return max_iter;
}
int main() {
    i64 scale = 1000, w = 800, h = 800, max_iter = 1000;
    i64 xmin = -2 * scale, xmax = 1 * scale, ymin = -1 * scale, ymax = 1 * scale;
    i64 total = 0;
    for (i64 y = 0; y < h; y++) {
        i64 ci = ymin + ((ymax - ymin) * y) / h;
        for (i64 x = 0; x < w; x++) {
            i64 cr = xmin + ((xmax - xmin) * x) / w;
            total += mandel(cr, ci, max_iter, scale);
        }
    }
    printf("total iterations = %ld\n", total);
    return 0;
}
